ST_resgroups.c

```diff
@@ -185,124 +181,124 @@ void rgroup_PartResources(void)
-
-	assert(g->max_age == 1);
-
-	if (!g->use_me)
-		return 0.0;
-
-	sumsize = 0.0;
-	ForEachGroupSpp(sp,rg,i)
-	{
-		s = Species[sp];
-		newsize = 0.0;
-		forced = FALSE;
-		// printf("add_seeds=%d \n",add_seeds);
-		if (!s->use_me)
-			continue;
-		//if add seeds is F and a random number drawn from the uniform distribution is less than prob of seed establishment
-		if (!add_seeds && RandUni() <= s->seedling_estab_prob)
-		{
-			/* force addition of new propagules */
-			//if regen_ok is TRUE, pass the g_pr parameter, if regen_ok is False, pass -1
-			_add_annual_seedprod(sp, (g->regen_ok) ? g_pr : -1.);
-			forced = TRUE;
-		}
-
-		//if regen_ok is T then x = values coming from _get_annual_maxestab, if regen ok is F then
-		//x=0.
-		x = (g->regen_ok) ? _get_annual_maxestab(sp) : 0.;
-		if (GT(x, 0.))
-		{
-			estabs = (x - (x / PR_0_est * g_pr)) * exp(-g_pr);
-			pr_inv = 1.0 / g_pr;
-			newsize = fmin(pr_inv, estabs);
-			if (add_seeds)
-			{
-				rgroup_AddSpecies(rg, sp);
-				Species_Update_Newsize(sp, newsize);
-			}
-		}
-
-		if (add_seeds && !forced)
-			_add_annual_seedprod(sp, (ZRO(x)) ? -1. : g_pr);
-
-		sumsize += newsize;
-	}
-
-	/*if (g->regen_ok == TRUE) {
-	 printf("Regen_ok %u\n",rg);}*/
-
-	/*will print out a flag of Regen_ok with the resourse group number (numbered
-	 * in the order of turned on groups in the "rgroup.in" file */
-
-	return (sumsize / (RealF) g->max_spp);
-
+static RealF _add_annuals(const GrpIndex rg, const SppIndex sp, const RealF lastyear_relsize) {
+    /*======================================================*/
+    /* check regen_ok flag.  if true, apply establishment and
+     * add to seedbank.  Otherwise, add 0 to seedbank and skip
+     * adding plants this year.  
+     * reset group size to account for additions, or 0 if none.
+     */
+	#define xF_DELTA (20*F_DELTA)
+	#define xD_DELTA (20*D_DELTA)
+	#define ZERO(x) \
+    ( (sizeof(x) == sizeof(float)) \
+  ? ((x)>-xF_DELTA && (x)<xF_DELTA) \
+  : ((x)>-xD_DELTA && (x)<xD_DELTA) )
+    IntU i, num_est; //number of individuals that will establish this year
+    RealF viable_seeds, viable_seeds_reduced; //viable seed and viable seed excluding this year's added seed, respectively
+    float alpha, beta, var; //parameters needed to calculation of num_est from beta distribution
+    GroupType *g;
+    SpeciesType *s;
+
+    g = RGroup[rg];
+    assert(g->max_age == 1);
+
+    s = Species[sp];
+    viable_seeds_reduced = 0.0;
+
+    /*Increment viable years for seeds and add seeds to the seedbank*/
+    _add_annual_seedprod(sp, lastyear_relsize);
+
+    /*Implement decay process and then get viable seeds from seed bank*/
+    viable_seeds = (g->regen_ok) ? _get_annual_maxestab(sp) : 0;
+
+    /*Create a beta random number draw, where mean =s->seedling_estab_prob
+     variance = s->var. 
+     Calculate alpha and beta based on the mean and variance values by species.in*/
+     alpha = (pow (s->seedling_estab_prob, 2) - pow (s->seedling_estab_prob, 3)
```

Comment:
And we should also check that `alpha >= 1 && beta >= 1` because we really aim to avoid a bimodal distribution with modes near 0 and 1.

Reply:
This was implemented by commit https://github.com/DrylandEcology/STEPWAT2/pull/107/commits/5e46b643bceabd4657a5bb61c07c63a3d7eb95bb